libs-base/Tools/gspath.1
.\"gspath(1) man page
.\"put together by Richard Frith-Macdoanld <rfm.gnu.org>
.\"Copyright (C) 2005 Free Software Foundation, Inc.
.\"Copying and distribution of this file, with or without modification,
.\"are permitted in any medium without royalty provided the copyright
.\"notice and this notice are preserved.
.\"
.\"Process this file with
.\"groff -man -Tascii gspath.1
.\"
.TH GSPATH 1 "July 2005" GNUstep "GNUstep System Manual"
.SH NAME
gspath \- Utility to display standard directory/path information
.SH SYNOPSIS
.B gspath
.P
.SH DESCRIPTION
This utility can be used to print out a single directory or path in order
for external applications to know what paths the GNUstep system is using.
It was written to provide information for the GNUstep make package.
.P
The
.B gspath
usility is called with exactly one argument, which determines what it prints.
.P
.SH Arguments
To display the path to the defaults directory for the current user, use the
.B defaults
argument.
.P
To display the directory search path needed to locate dynamic libraries,
use the
.B libpath
argument.
.P
To display the directory search path needed to locate commmand-line programs,
use the
.B path
argument.
.P
To display the GNUstep home directory of the current user, use the
.B user
argument.
.P
.SH HISTORY
Work on
.B gspath
started July 2005.
.PP
This manual page first appeared in gnustep-base 1.11.1 (July 2005).
.P
.SH AUTHORS
.B gspath 
was written by Richard Frith-Macdonald <rfm@gnu.org>.
.PP
This man page was written by Richard Frith-Macdonald <rfm@gnu.org>.


libs-gui/NEWS
1 NEWS
******

The currently released version of the library is '0.28.0'.

1.1 Noteworthy changes in version '0.28.0'
==========================================

This version adds support for modern XIB files and many new classes.
Plus the usual bunch of bug fixes.

   * Support loading of document XIB files.
   * Improve Key Value Binding for NSArrayController and add more
     bindings.
   * Better support for multi monitor usage and other improvement in the
     backend integration.
   * Add classes NSFontCollection, NSColorSampler, NSSpeechRecognizer,
     NSAppearance, NSPDFInfo, NSPICTImageRep, NSCIImageRep,
     NSPDFImageRep, NSPDFPanel, NSDataAsset, NSDatePicker,
     NSDatePickerCell, NSPredicateEditor, NSPredicateEditorRowTemplate,
     NSRuleEditor, NSGestureRecognizer, NSButtonTouchBarItem,
     NSCandidateListTouchBarItem, NSClickGestureRecognizer,
     NSColorPickerTouchBarItem, NSCustomTouchBarItem,
     NSGroupTouchBarItem, NSMagnificationGestureRecognizer,
     NSPanGestureRecognizer, NSPickerTouchBarItem,
     NSPopoverTouchBarItem, NSPressGestureRecognizer,
     NSRotationGestureRecognizer, NSSharingServicePickerTouchBarItem,
     NSSliderTouchBarItem, NSStepperTouchBarItem, NSTouchBarItem,
     NSTouchBar, NSTouch, NSDockTile.
   * Implement NSEPSImageRep.
   * Better encoding handling in RTF files.
   * Theming and drawing improvements.
   * Increase small font size to 10.
   * New cursor and stepper images.
   * Move NSFileWrapper to Foundation.
   * Fixed build on Debian GNU/kFreeBSD.
   * With command line argument -autolaunch YES, do not activate the
     application when -activateIgnoringOtherApps: is invoked.
   * Improvements to WindowMaker compatibility (e.g.  WMFHideApplication
     support).
   * Lowered NSFloatingWindowLevel by one to distinguish floating panels
     from menus.

1.2 Noteworthy changes in version '0.27.0'
==========================================

This version includes numerous bugfixes, compatibility improvements and
other changes accumulated over the last year.  It also enables work to
be done on integrating NSViews with a Core Animation renderer.

   * Make targetForAction safer.
   * Speed up menu updates.
   * Clean up speech tool compilation and switch to newer interface.
   * Fix bug in CUPS subclassing introduced in last release.
   * Minor improvements to typesetting.
   * Add NSIsControllerMarker.
   * Fix tracking on segmented cell.
   * Bring slider cell closer to Cocoa implementation.
   * Add ivar for Core Animation in NSView.
   * Improve border calculation on printing.
   * Lazy load app icon.
   * Better detection of removable volumes.
   * Polish translations.
   * Japanese translations.
   * Lots of bug fixes.

1.3 Noteworthy changes in version '0.26.2'
==========================================

This version is a small, but important bugfix release.

   * printing: Fix allocation of the CUPS printing classes.
   * installation: Fix the configure script.

1.4 Noteworthy changes in version '0.26.1'
==========================================

This version is released to conincide with version 1.25.1 of
gnustep-base, which contains changes required for this version of
gnustep-gui and gnustep-back.

   It includes an important workaround for users of GNUstep Objective-C
Runtime (libobjc2) and non-fragile ABI to avoid a bug in interaction
between the clang compiler and the runtime when non-fragile ABI is in
use.  Specifically, Clang and the runtime may disagree on what is the
offset of an ivar in a class's RAM. This manifested in a crash at
application startup due to misalignment of _gcontext inside NSThread.
See the mailing list discussion
(http://lists.gnu.org/archive/html/discuss-gnustep/2017-12/msg00129.html)
for more information.

   It also contains the following changes:

   * tests: Cleanup of warnings.
   * tests: Fix text system deallocation test.
   * printing: Undefine __BLOCKS__ before including cups.h, as some
     versions of the header expect that libdispatch is present and used
     if __BLOCKS__ is defined.
   * graphics context: Workaround for Clang+libobjc2+nonfragile ABI
     issue.

1.5 Noteworthy changes in version '0.26.0'
==========================================

This version was bumped due to previous binary incompatibilities between
0.25.0 and 0.25.1.  This version also includes numerous compatibility
improvements from the Summer of Code project, and a wide variety of
other fixes.  Notably, it fixes the use of cupsGetPPD() in the printing
system.

   * printing: Add an include to get deprecated function cupsGetPPD() on
     newer CUPS systems.
   * chore: Bump required base version.
   * tiff: Support for writing resolution.
   * jpeg: Save resolution information if it is different from 72 dpi.
   * save panel: Fix return type of sorting function.
   * events: Add some newer Cocoa enums and one method with dummy
     implementation.
   * speech synthesis: NSSpeechSynthesizerDelegate is now a @protocol on
     runtimes that support it.
   * pasteboard: New type identifiers.
   * translations: Some work on Polish, Russian and German translations
   * cell: Improvements to mouse tracking logic on NSCell.
   * image: If an unknown named image is unarchived with a coder or
     keyed coder, keep the name.
   * screen: Add -backingScaleFactor and return 1.0.
   * window: Return 1.0 from -backingScaleFactor.

   * compatibility: Numerous stub implementations of constants, classes
     and methods to improve source-level compatibility.
   * other bugfixes

1.6 Noteworthy changes in version '0.25.1'
==========================================

   * JPEG (saving) alpha channel fixes and size with resolution != 72
   * JPEG resolution read support
   * TIFF saving fixes
   * Improved volumes mounting and support
   * Portability improvements in volume mounting and support
   * Corrected layout of empty strings
   * Only update visible menus

1.7 Noteworthy changes in version '0.25.0'

libs-base/Tools/gspath.m
/** This utility provides path/directory layout information for GNUstep.
   Copyright (C) 2005 Free Software Foundation, Inc.

   Written by:  Richard Frith-Macdonald <richard@brainstorm.co.uk>
   Created: July 2005

   This file is part of the GNUstep Project

   This program is free software; you can redistribute it and/or
   modify it under the terms of the GNU General Public License
   as published by the Free Software Foundation; either
   version 3 of the License, or (at your option) any later version.

   You should have received a copy of the GNU General Public
   License along with this program; see the file COPYINGv3.
   If not, write to the Free Software Foundation,
   51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.

   */

#import "common.h"

#import	"Foundation/NSArray.h"
#import	"Foundation/NSAutoreleasePool.h"
#import	"Foundation/NSPathUtilities.h"
#import	"Foundation/NSProcessInfo.h"
#import	"Foundation/NSUserDefaults.h"


/**
 <p>The 'gspath' utility prints out various items of path/directory
 information (one item at a time).<br />
 The program always takes a single argument ... selecting the information
 to be printed.</p>
 The arguments and their meanings are -<br />
 <deflist>
   <term>defaults</term>
   <desc>The GNUstep defaults directory of the current user</desc>
   <term>devpath</term>
   <desc>A path specification which may be used to add the root(s) of
     the GNUstep development environment on the current system.</desc>
   <term>libpath</term>
   <desc>A path specification which may be used to add all the standard
     GNUstep directories where dynamic libraries are normally stored.<br />
     you might do 'LD_LIBRARY_PATH=$LD_LIBRARY_PATH:`gspath libpath`' to make
     use of this.</desc>
   <term>path</term>
   <desc>A path specification which may be used to add all the standard
     GNUstep directories where command-line programs are normally stored.<br />
     you might do 'PATH=$PATH:`gspath path`' to make use of this.</desc>
   <term>user</term>
   <desc>The GNUstep home directory of the current user</desc>
 </deflist>
 */
int
main(int argc, char** argv, char **env)
{
  NSAutoreleasePool	*pool;
  NSProcessInfo		*proc;
  NSArray		*args;

#ifdef GS_PASS_ARGUMENTS
  GSInitializeProcess(argc, argv, env);
#endif
  pool = [NSAutoreleasePool new];
  proc = [NSProcessInfo processInfo];
  if (proc == nil)
    {
      GSPrintf(stderr, @"gspath: unable to get process information!\n");
      [pool drain];
      return 1;
    }

  args = [proc arguments];

  if ([args count] == 2)
    {
      BOOL	ok = YES;
      NSString	*name = [[args objectAtIndex: 1] lowercaseString];
      NSString	*sep;

#ifdef	__MINGW__
      sep = @";";
#else
      sep = @":";
#endif

      if ([name isEqualToString: @"defaults"] == YES)
	{
	  GSPrintf(stdout, @"%@", GSDefaultsRootForUser(nil));
	}
      else if ([name isEqualToString: @"path"] == YES)
	{
	  NSArray	*directories;
	  NSString	*path;

	  directories = NSSearchPathForDirectoriesInDomains
	    (GSToolsDirectory, NSAllDomainsMask, YES);
	  path = [directories componentsJoinedByString: sep];
	  GSPrintf(stdout, @"%@", path);
	}
      else if ([name isEqualToString: @"devpath"] == YES)
	{
	  NSArray	*directories;
	  NSString	*path;

	  directories = NSSearchPathForDirectoriesInDomains
	    (NSDeveloperDirectory, NSAllDomainsMask, YES);
	  path = [directories componentsJoinedByString: sep];
	  GSPrintf(stdout, @"%@", path);
	}
      else if ([name isEqualToString: @"libpath"] == YES)
	{
	  NSArray	*directories;
	  NSString	*path;

	  directories = NSSearchPathForDirectoriesInDomains
	    (GSLibrariesDirectory, NSAllDomainsMask, YES);
	  path = [directories componentsJoinedByString: sep];
	  GSPrintf(stdout, @"%@", path);
	}
      else if ([name isEqualToString: @"user"] == YES)
	{
	  GSPrintf(stdout, @"%@", NSHomeDirectory());
	}
      else
	{
	  ok = NO;	// Unrecognised option
	}
      if (ok == YES)
	{
	  [pool drain];
	  return 0;
	}
    }

  GSPrintf(stderr,
@"The 'gspath' utility prints out various items of path/directory\n"
@"information (one item at a time).\n"
@"The program always takes a single argument ... selecting the information\n"
@"to be printed.\n\n"
@"The arguments and their meanings are -\n\n"
@"defaults\n"
@"  The GNUstep defaults directory of the current user.\n\n"
@"devpath\n"
@"  A path specification which may be used to add the root(s) of\n"
@"  the GNUstep development environment on the current system.\n\n"
@"libpath\n"
@"  A path specification which may be used to add all the standard GNUstep\n"
@"  directories where dynamic libraries are normally stored.\n\n"
@"  you might do 'LD_LIBRARY_PATH=$LD_LIBRARY_PATH:`gspath libpath`' to make\n"
@"  use of this.\n\n"
@"path\n"
@"  A path specification which may be used to add all the standard GNUstep\n"
@"  directories where command-line programs are normally stored.\n"
@"  you might do 'PATH=$PATH:`gspath path`' to make use of this.\n\n"
@"user\n"
@"  The GNUstep home directory of the current user\n\n"
);
  [pool drain];
  return 1;
}
